pkg/bgpv1/manager/reconcilerv2/neighbor.go

```diff
@@ -0,0 +1,362 @@
+	peerConfig store.BGPCPResourceStore[*v2alpha1.CiliumBGPPeerConfig],
+	daemonConfig *option.DaemonConfig) NeighborReconcilerOut {
+
+	return NeighborReconcilerOut{
+		Reconciler: &NeighborReconciler{
+			SecretStore:  secretStore,
+			PeerConfig:   peerConfig,
+			DaemonConfig: daemonConfig,
+		},
+	}
+}
+
+// PeerData keeps a peer and its configuration. It also keeps the TCP password from secret store.
+type PeerData struct {
+	Peer     *v2alpha1.CiliumBGPNodePeer
+	Config   *v2alpha1.CiliumBGPPeerConfigSpec
+	Password string
+}
+
+func (p *PeerData) Equal(other *PeerData) bool {
+	if p == nil && other == nil {
+		return true
+	}
+
+	if (p == nil && other != nil) || (p != nil && other == nil) {
+		return false
+	}
+
+	if p.Peer.DeepEqual(other.Peer) && p.Config.DeepEqual(other.Config) && p.Password == other.Password {
+		return true
+	}
+
+	return false
+}
+
+// NeighborReconcilerMetadata keeps a map of running peers to peer configuration.
+// key is the peer name.
+type NeighborReconcilerMetadata map[string][]*PeerData
+
+func (r *NeighborReconciler) getMetadata(i *instance.BGPInstance) NeighborReconcilerMetadata {
+	if _, found := i.Metadata[r.Name()]; !found {
+		i.Metadata[r.Name()] = make(NeighborReconcilerMetadata)
+	}
+	return i.Metadata[r.Name()].(NeighborReconcilerMetadata)
+}
+
+func (r *NeighborReconciler) upsertMetadata(i *instance.BGPInstance, instanceName string, d *PeerData) {
+	if i == nil || d == nil {
+		return
+	}
+
+	neighMetadata := r.getMetadata(i)
+
+	peers, exists := neighMetadata[instanceName]
+	if !exists {
+		neighMetadata[instanceName] = []*PeerData{d}
+		return
+	}
+
+	found := false
+	for i, p := range peers {
+		if p.Peer.Name == d.Peer.Name {
+			peers[i] = d
+			found = true
+			break
+		}
+	}
+
+	if !found {
+		peers = append(peers, d)
+	}
+
+	neighMetadata[instanceName] = peers
+}
+
+func (r *NeighborReconciler) deleteMetadata(i *instance.BGPInstance, instanceName string, d *PeerData) {
+	if i == nil || d == nil {
+		return
+	}
+
+	neighMetadata := r.getMetadata(i)
+	peers, exists := neighMetadata[instanceName]
+	if !exists {
+		return
+	}
+
+	for i, p := range peers {
+		if p.Peer.Name == d.Peer.Name {
+			peers[i] = peers[len(peers)-1]
+			peers = peers[:len(peers)-1]
+			break
+		}
+	}
+
+	neighMetadata[instanceName] = peers
+}
+
+func (r *NeighborReconciler) Name() string {
+	return "Neighbor"
+}
+
+// Priority of neighbor reconciler is higher than pod/service announcements.
+// This is important for graceful restart case, where all expected routes are pushed
+// into gobgp RIB before neighbors are added. So, gobgp can send out all prefixes
+// within initial update message exchange with neighbors before sending EOR marker.
+func (r *NeighborReconciler) Priority() int {
+	return 60
+}
+
+func (r *NeighborReconciler) Reconcile(ctx context.Context, p ReconcileParams) error {
+	if p.DesiredConfig == nil {
+		return fmt.Errorf("attempted neighbor reconciliation with nil CiliumBGPNodeInstance")
```

Comment:
These errors are mostly for developers to debug problem. I don't think we should run into this condition, it is more of pragmatic approach, that if we have this corner case - we should error out instead of panic at later stage due to unexpected request. I think having internal details in error message is useful for doing RCA. From user prospective, if they see this error - they should probably just open an issue.

Reply:
Ok, then how about prefixing the error with `BUG:`? So that we can encourage users to file an issue. Same for below.

Reply:
My concern with adding 'bug' in error message is it's not scalable. Most of the errors are a problem, most likely a bug. Logging error imo itself an indication of a problem.

Reply:
> Most of the errors are a problem, most likely a bug

Error and bug are two different things IMO.

For example, [here](https://github.com/cilium/cilium/pull/30108/files#diff-23b500b5d2af6a22a607b920886922d961deb0e5f8c686f7481e4320722e017fR646) we return an error when the node annotation has a format error. We log it as an Error later. However, this indicates the problem of the user's configuration and not a bug. Our code has no problem, but the user's configuration has a problem. This is an example of `Error`.

On the other hand, in this case, our code is likely to have a bug when we hit this condition. Users have nothing to do; we know the only way to solve this is by modifying the code. Then, we want them to report an issue. This is an example of `Bug`.

Both cases are currently reported as an error-level log. Thus, users don't have any way to distinguish those two. That's why I'm suggesting to indicate it is a `BUG` explicitly.

Reply:
I can see the `BUG:` prefix pattern in some other cilium packages as well, so in general I wouldn't be opposed it.
To me, it makes sense in logs, but it appears a bit strange in error strings - which often can be wrapped in another error, so the `BUG:` prefix may eventually appear in the middle of the log line when eventually logged.

What about logging with `BUG:` prefix at this place, and returning an error without the prefix?

Reply:
In these particular cases, we are logging error message so I just prefixed the error message with `BUG:`. I am not fully opposed to the idea as it can be context dependent and maybe help users to raise issue by looking at the errors. 

( feel free to ignore rest of the comment )

My counter argument is 'error' level log is enough to warrant investigation. If error is result of a bug or incorrect configuration, it is impossible to tell without looking more into the context around it. 

In the example of node annotation - User sending incorrectly formatted annotation is user config issue, however if we introduced a bug by modifying annotation parsing, it is not a user error. In which case, we have to investigate the logged error message. 

From logging prospective, the obvious cases which should never ever happen should result in fatal. But that is bit extreme as well. 

I guess, main intention here is to get attention of the user by logging 'BUG' in the error message. My main point is that every error message should get a red-flag and warrants the investigation. Initially by user to validate that configuration is correct and then an issue to get extra eyes on it.

Anyway, it's an interesting thought experiment. I prefixed the logs with BUG in this case :)